sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/CbsTokenProvider.cs

```diff
@@ -62,14 +83,94 @@ internal sealed class CbsTokenProvider : ICbsTokenProvider
+        /// </summary>
+        ///
+        /// <returns>The <see cref="CbsToken"/> to use for authorization.</returns>
+        ///
+        private async Task<CbsToken> AcquireJwtTokenAsync()
+        {
+            var token = Volatile.Read(ref _cachedJwtToken);
+
+            // If the token is expiring, attempt to acquire the semaphore needed to
+            // refresh it.
+
+            if (IsExpiring(token))
+            {
+                try
+                {
+                    // Attempt to acquire the semaphore synchronously, in case it is not currently
+                    // held.
+
+                    if (!TokenAcquireSemaphore.Wait(0, CancellationToken.None))
+                    {
+                        await TokenAcquireSemaphore.WaitAsync(CancellationToken).ConfigureAwait(false);
+                    }
+
+                    // Because the token may have been refreshed while waiting for the semaphore,
+                    // check expiration again to avoid making an necessary request.  Because the token
+                    // is only updated when the semaphore is held, there's no need to perform a volatile
+                    // read for this refresh.
+
+                    token = _cachedJwtToken;
+
+                    if (IsExpiring(token))
+                    {
+                        var accessToken = await Credential.GetTokenUsingDefaultScopeAsync(CancellationToken).ConfigureAwait(false);
+                        token = new CbsToken(accessToken.Token, TokenType, accessToken.ExpiresOn.UtcDateTime);
+
+                        _cachedJwtToken = token;
+                    }
+                }
+                finally
+                {
+                    TokenAcquireSemaphore.Release();
+                }
+            }
+
+            return token;
+        }
+
+        /// <summary>
+        ///   Acquires a token for a shared access credential.
+        /// </summary>
+        ///
+        /// <returns>The <see cref="CbsToken"/> to use for authorization.</returns>
+        ///
+        private async Task<CbsToken> AcquireSharedAccessTokenAsync()
         {
-            AccessToken token = await Credential.GetTokenUsingDefaultScopeAsync(CancellationToken).ConfigureAwait(false);
-            return new CbsToken(token.Token, TokenType, token.ExpiresOn.UtcDateTime);
+            var accessToken = await Credential.GetTokenUsingDefaultScopeAsync(CancellationToken).ConfigureAwait(false);
+            return new CbsToken(accessToken.Token, TokenType, accessToken.ExpiresOn.UtcDateTime);
         }
+
+        /// <summary>
+        ///   Determines whether the specified token is considered to be expiring.
+        /// </summary>
+        ///
+        /// <param name="token">The token to consider.</param>
+        ///
+        /// <returns><c>true</c> if the specified token is expiring; otherwise, <c>false</c>.</returns>
+        ///
+        private bool IsExpiring(CbsToken token) =>
```

Comment:
nit: perhaps this method could be called something like `NeedsRefresh` or `IsNearExpiration` since really all tokens are technically expiring.

Reply:
Good thought.... I especially like the second option and will steal that.